src/server/main_service.cc

```diff
@@ -68,6 +68,118 @@ std::optional<VarzFunction> engine_varz;
+        absl::StrAppend(&result, "\\\"");
+        break;
+      case '\n':
+        absl::StrAppend(&result, "\\n");
+        break;
+      case '\r':
+        absl::StrAppend(&result, "\\r");
+        break;
+      case '\t':
+        absl::StrAppend(&result, "\\t");
+        break;
+      case '\a':
+        absl::StrAppend(&result, "\\a");
+        break;
+      case '\b':
+        absl::StrAppend(&result, "\\b");
+        break;
+      default:
+        if (isprint(c)) {
+          result += c;
+        } else {
+          absl::StrAppend(&result, "\\x", absl::Hex((unsigned char)c, absl::kZeroPad2));
+        }
+        break;
+    }
+  }
+  absl::StrAppend(&result, "\"");
+  return result;
+}
+
+std::string MakeMonitorMessage(const ConnectionState& conn_state,
+                               const facade::Connection* connection, CmdArgList args) {
+  std::string message = MonitorTimestamp();
+
+  auto AppendConnDetails = [&]() {
+    if (conn_state.script_info.has_value()) {
+      absl::StrAppend(&message, "lua] ");
+    } else {
+      absl::StrAppend(&message, connection->RemoteEndpointStr());
+    }
+  };
+  auto AppendCmd = [&]() {
+    if (args.empty()) {
+      absl::StrAppend(&message, "error - empty cmd list!");
+    } else if (auto cmd_name = std::string_view(args[0].data(), args[0].size());
+               cmd_name == "AUTH") {  // we cannot just send auth details in this case
+      absl::StrAppend(&message, "\"", cmd_name, "\"");
+    } else {
+      message = std::accumulate(args.begin(), args.end(), message, [](auto str, const auto& cmd) {
+        absl::StrAppend(&str, " ",
+                        CmdEntryToMonitorFormat(std::string_view(cmd.data(), cmd.size())));
+        return str;
+      });
+    }
+  };
+
+  AppendConnDetails();
+  AppendCmd();
+  return message;
+}
+
+void MonitorCmd(bool admin_cmd, ConnectionContext* connection, CmdArgList args) {
+  // We are not sending any admin command in the monitor, and we do not want to
+  // do any processing if we don't have any waiting connections with monitor
+  // enabled on them - see https://redis.io/commands/monitor/
+  const auto& my_monitors = ServerState::tlocal()->Monitors();
+  if (!(my_monitors.Empty() || admin_cmd)) {
+    //  We have connections waiting to get the info on the last command, send it to them
+    auto monitor_msg = MakeMonitorMessage(connection->conn_state, connection->owner(), args);
+    // Note that this is accepted by value for lifetime reasons
+    // we want to have our own copy since we are assuming that
+    // 1. there will be not to many connections that we in monitor state
+    // 2. we need to have for each of them each own copy for thread safe reasons
+    VLOG(1) << "sending command '" << monitor_msg << "' to the clients that registered on it";
+    shard_set->pool()->Await([monitor_msg](unsigned idx, util::ProactorBase*) {
+      ServerState::tlocal()->Monitors().Send(monitor_msg, idx);
+    });
+
+    shard_set->pool()->Await(
```

Comment:
Now, I am not sure I understand why we need releasing tokens. Probably missing something.

Reply:
This token is about the connection, we need to notify that we are done with it after we finish sending. This is the same logic we have for these tokens (not the message token) in the pubsub. Since we are sharing the connection, and the connection may be closed while we still have the message on the "way" to the socket, we need to make sure that its will stay alive. Once the messages were sent, we  can notify that we done with this.

Reply:
I do not think it's the same here. The dangerous points are where we make inter-threads hops  or can block.
when you send a message, you first send it to destination thread (not dangerous) and only then go over connections inside thread-local storage (again, not dangerous). so what scenario is dangerous?

Reply:
The token was removed from the code:
1. We are running the send from the context of the thread, each thread has its own connections only.
2. The data that the connection is sending is "private" we no longer need to sync on it.
3. If a connection is to be removed, it will happen before we are using any connection, since the monitor operation is done only at the end of the processing from incoming new command. So there no risk of race. The cleanup for any connection that is not longer needed is done before sending and the connection that is monitor state can only used for monitor (it cannot process any other command - other then closing the connection).